### DISCUSSION TITLE
`optimization.concatenateModules` leading to nondeterministic build output

### DISCUSSION THREAD

**Discussion by pgoldberg** [webpack/webpack#15844] [Q&A]
I recently upgraded to webpack 5, and I'm noticing that our builds no longer have deterministic outputs. I've narrowed down the issue to `optimization.concatenateModules` – when I set it to `false`, I get the same output on every build, although I'd like to keep `concatenateModules` on. With it on, the output just differs slightly. It looks like the modules are being concatenated in different orders? For example:

```
-        Et = new pt(Ct),
-        kt = new ht(Ct);
+        Et = new ht(Ct),
+        kt = new pt(Ct);
```

Any idea what might be causing this? Not sure if it's a bug in the `ModuleConcatenationPlugin`, but given that I haven't seen any other issues about this or any mention of this problem online, I figure maybe there's something in our configuration that could be causing this?

**Comment by alexander-akait** [comment]
Do you have `ModuleConcatenationPlugin` in `plugins` or use `optimization.concatenateModules` to setup it?

**Reply by pgoldberg** [reply]
We were just using the defaults (`optimization.concatenateModules` on in production mode)

**Reply by alexander-akait** [reply]
Can you provide list of plugins?

**Reply by pgoldberg** [reply]
Here's a list of the plugins we use in production:
```
"@cerner/duplicate-package-checker-webpack-plugin"
"circular-dependency-plugin"
"css-minimizer-webpack-plugin"
"mini-css-extract-plugin"
"webpack-retry-chunk-load-plugin"
"copy-webpack-plugin"
"html-webpack-plugin"
"monaco-editor-webpack-plugin"
"webpack-bundle-analyzer"
"webpack-stats-plugin"
webpack.DefinePlugin
webpack.EnvironmentPlugin
webpack.ProvidePlugin
```

I've gone through and disabled almost all of these one by one, but none of these fixed the issue

**Reply by alexander-akait** [reply]
Previously I see the same problem with `monaco-editor-webpack-plugin`, can you try to disable it and check?

**Reply by pgoldberg** [reply]
Just tried removing it again, but no luck 😕

**Reply by alexander-akait** [reply]
Can you try to disable step by step plugins and check it? I think some plugin change something in chunks so you get another name in each build

**Reply by pgoldberg** [reply]
I've gone through and removed every plugin for testing and I'm still consistently seeing the output differ between two separate builds. I've also gone through and deleted a lot of things from the config to see if something else might be causing it, and I'm still seeing the issue. The only way I've been able to get it to go away is by disabling `concatenateModules`.

**Reply by alexander-akait** [reply]
@pgoldberg sounds like a bug, can you try to create minimum reproducible test repo?

**Reply by pgoldberg** [reply]
Hey @alexander-akait,

I struggled to get a minimal repro together, since I'm not totally sure what causes webpack to run into this issue, but I think I was able to narrow down where the issue is popping up. Here in `ConcatenatedModule`: https://github.com/webpack/webpack/blob/2738eebc7880835d88c727d364ad37f3ec557593/lib/optimize/ConcatenatedModule.js#L905-L907

It seems like for some of the `references` that have the same `sourceOrder`, falling back to the original order in the `references` array is nondeterministic. When I changed that line to this (very hacky/gross, just trying to get something that would be deterministic so I could try to narrow down the issue), the problem went away:
```
references.sort(
	concatComparators(
		bySourceOrder,
		(a, b) =>
			a.connection.dependency.loc.start.line < b.connection.dependency.loc.start.line
				? -1
				: b.connection.dependency.loc.start.line < a.connection.dependency.loc.start.line
				? 1
				: a.connection.dependency.loc.start.column < b.connection.dependency.loc.start.column
				? -1
				: b.connection.dependency.loc.start.column < a.connection.dependency.loc.start.column
				? 1
				: 0,
		keepOriginalOrder(references),
	),
);
```

**Reply by alexander-akait** [reply]
Can you look which dependecy affected it? i.e. a and b when it is unstable

**Reply by pgoldberg** [reply]
It happens with several of our dependencies, but they all seem to be our internal dependencies (not internal to this particular project, but internal to our company). Is there anything in particular I could look for in the dependencies that might be useful information? I unfortunately can't share the code of those packages

**Reply by alexander-akait** [reply]
There are two things:
1. Some of plugins change dependecies so you have this problem
2. We really have a bug and in this case we need investigate it

Maybe you can provide output from `bySourceOrder`?

**Reply by vankop** [reply]
Could you provide webpack.config?

**Reply by pgoldberg** [reply]
It's definitely a bug – I'm getting the issue without any plugins added. I was able to get a minimal repro together here: https://github.com/pgoldberg/concatenate-modules-bug.

It seems that there's an issue related to re-exporting within a package's `index.js`.

**Reply by pgoldberg** [reply]
I can file an issue if that's easier for tracking!

**Reply by pgoldberg** [reply]
After looking through a tiny bit more, I think the bug might be upstream of the `ModuleConcatenationPlugin` in the `SideEffectsFlagPlugin`. My guess is somewhere in here: https://github.com/webpack/webpack/blob/2738eebc7880835d88c727d364ad37f3ec557593/lib/optimize/SideEffectsFlagPlugin.js#L243-L314

Turning `optimization.sideEffects` off also seems to stop the bug – also seems to point to the problem being somewhere in the side effects logic

**Reply by vankop** [reply]
I'll take a look

**Reply by pgoldberg** [reply]
Thank you! I looked a little closer and it seems like the `SideEffectsFlagPlugin` expects [the `modules` received from the `optimizeDependencies` compilation hook](https://github.com/webpack/webpack/blob/2738eebc7880835d88c727d364ad37f3ec557593/lib/optimize/SideEffectsFlagPlugin.js#L237) to have a deterministic ordering, but the order seems to change in the minimal repro. I dug a bit deeper and got a little lost, but I saw that `onDependenciesSorted` can be called in a different order here, which might be the root of the problem: https://github.com/webpack/webpack/blob/2738eebc7880835d88c727d364ad37f3ec557593/lib/Compilation.js#L1463. Seems like it's maybe some kind of race condition?

**Reply by pgoldberg** [reply]
Looking a little more, it seems like `handleModuleCreation` is always called in the correct order for the dependencies of `dep/index.js` ([here](https://github.com/webpack/webpack/blob/2738eebc7880835d88c727d364ad37f3ec557593/lib/Compilation.js#L1476)), but then the callbacks for `factorizeModule` on those dependencies sometimes get called in a different order ([here](https://github.com/webpack/webpack/blob/2738eebc7880835d88c727d364ad37f3ec557593/lib/Compilation.js#L1774-L1880)), which then causes `inner_dep2` to get processed before `inner_dep`.

`dep/index.js`, `inner_dep`, and `inner_dep2` are all referring to the minimal repro, for context

**Reply by alexander-akait** [reply]
@pgoldberg Can you try to set https://webpack.js.org/configuration/other-options/#parallelism to `1` and check? If you think it is race it should help, and it means we have a bug somewhere...

**Reply by pgoldberg** [reply]
Just did it with the minimal repro and it does seem like the issue goes away – looks like it must be a race condition. After spending some time debugging with the test repo, it seems like it's a race condition with `NormalModuleFactory.create` for `inner_dep2/index.js` finishing before `inner_dep/index.js` here: https://github.com/webpack/webpack/blob/2738eebc7880835d88c727d364ad37f3ec557593/lib/Compilation.js#L1973-L2033

That's the first place I see the order become incorrect

**Reply by alexander-akait** [reply]
So let's open an issue fir this, sounds like a bug

**Reply by pgoldberg** [reply]
Filed https://github.com/webpack/webpack/issues/15866

**Reply by vankop** [reply]
factorizing in different order is ok since modules itself are independent, but this should not lead to different concat results

**Reply by vankop** [reply]
probably this happens because as a non-determenistic result of `compilation.modules` ( Set<Module> )
